pkg/webui/oauth/views/update-password/index.js

```diff
@@ -0,0 +1,230 @@
+    .matches(uppercase, sharedMessages.validatePasswordUppercase)
+    .matches(special, sharedMessages.validatePasswordSpecial)
+    .required(sharedMessages.validateRequired),
+  confirm: Yup.string()
+    .oneOf([ Yup.ref('password'), null ], sharedMessages.validatePasswordMatch)
+    .min(8)
+    .required(sharedMessages.validateRequired),
+})
+
+const initialValues = {
+  password: '',
+  confirm: '',
+  password_changed: false,
+  revoke_all_access: true,
+}
+
+@connect(state => ({
+  fetching: state.user.fetching,
+  user: state.user.user,
+}), {
+  handleCancelUpdate: () => push('/'),
+  push,
+})
+@bind
+export default class UpdatePassword extends React.PureComponent {
+
+  static propTypes = {
+    fetching: PropTypes.bool.isRequired,
+    user: PropTypes.object,
+    handleCancelUpdate: PropTypes.func.isRequired,
+    push: PropTypes.func.isRequired,
+  }
+
+  constructor (props) {
+    super(props)
+
+    this.state = {
+      error: '',
+      info: '',
+      revoke_all_access: true,
+    }
+  }
+
+  handleRevokeAllAccess (evt) {
+    this.setState({ revoke_all_access: evt.target.checked })
+  }
+
+  async handleSubmit (values, { resetForm, setSubmitting }) {
+    const { user, push } = this.props
+    const userParams = queryString.parse(this.props.location.search)
+    const oldPassword = values.old_password ? values.old_password : userParams.current
+    const userId = Boolean(user) ? user.ids.user_id : userParams.user
+    try {
+      await api.users.updatePassword( userId, {
+        user_ids: { user_id: userId },
+        'new': values.password,
+        old: oldPassword,
+        revoke_all_access: values.revoke_all_access,
+      })
+      if (Boolean(user)) {
+        this.setState({
+          error: '',
+          info: m.passwordChanged,
+          changed: true,
+        })
+        resetForm(initialValues)
+      } else {
+        push('/login', {
+          info: m.passwordChanged,
+        })
+      }
+    } catch (error) {
+      this.setState({
+        error: error.response.data,
+        info: '',
+      })
+      setSubmitting(false)
+    }
+  }
+
+  handleCancel () {
+    this.props.handleCancelUpdate()
+  }
+
+  render () {
+    const {
+      user,
+      fetching,
+    } = this.props
+
+    const {
+      error,
+      info,
+      revoke_all_access,
+    } = this.state
+
+    const userParams = queryString.parse(this.props.location.search)
+
+    let oldPasswordField
+
+    if (fetching) {
+      return (
+        <Spinner center>
+          <Message content={sharedMessages.loading} />
+        </Spinner>
+      )
+    }
+
+    if (Boolean(user)) {
+      oldPasswordField = (
+        <Form.Field
+          component={Input}
+          required
+          title={m.oldPassword}
+          name="old_password"
+          type="password"
+          autoComplete="old-password"
+          autoFocus
+        />
+      )
+    } else if (!userParams.user || !userParams.current) {
```

Comment:
this doesn't make sense, this is a guard for accessing the update password view when not logged in and without query parameters. that's why it first checks if the user object is present, and if it isn't it checks for query parameters. If it would first check for query parameters then it would always return redirect even when the user is logged in since the query parameters are missing.

Reply:
https://gist.github.com/bafonins/13ef21b491073a65c6e23d8e911e26b0 ?